### ISSUE TITLE
New calling convention for Pallene entry point

### ISSUE THREAD

**Issue by hugomg** [pallene-lang/pallene#413]
# Objectives

This issue is about the planned cleanup of the `G` and the new calling convention that would come together with it. The main objectives are:

1. Treat global variables and module functions as regular upvalues
    * They are only an upvalue of closures that actually use them, to avoid memory leaks.
    * We want to remove the special treatment of global vars in the IR and Coder.
2. Potentially speed up calls to functions with many upvalues. (We should measure this)

# The calling convention

The example uses the inner lambda in the following program:

```lua
local function f(): (integer)->integer
    local x = 1
    local y = 2
    return function(z, w)
        return x + y + z + w
    end
end
```

The old calling convention puts all the global stuff in a `G` userdata and passes each upvalue as a separate argument.

```c
// Old calling convention
static lua_Integer function_05(
    lua_State *L,     /* Lua state */
    Udata *G,         /* module variables, strings & metatables*/
    StackValue *base, /* Lua stack*/
    lua_Integer uv1,  /* upvalues*/
    lua_Integer uv2, 
    lua_Integer x1,   /* arguments */
    lua_Integer x2 
) {
```

I propose putting only the strings and metatables in the `G`, which will now be called the constant table  `K`. The global variables will become regular upvalues. Instead of having one function argument per upvalue, the Pallene entry point will receive a pointer to the upvalue array `U`. 

The reason we are currently using separate arguments for the upvalues was in the hope that we might not need to allocate a closure object in the first place. However, I 'm not sure if we really want to avoid these closures allocations... I also worry that all those extra arguments might spill into the C stack. Passing the upvalue array in a single argument might be simpler and more efficient.

```c
// New calling convention
static lua_Integer function_05(
    lua_State *L,     /* Lua state */
    StackValue *base, /* Lua stack*/
    TValue *K,        /* array of module strings & metatables */
    TValue *U,        /* array of upvalues */
    lua_Integer x1,   /* arguments */
    lua_Integer x2
) {
```

# TValue\*

The constant table argument may either be an `Udata *` (as today) or it may be a `TValue *` that points to the upvalue array of the userdata.

Similarly, the upvalue argument may be either a `CClosure *` or it may be a `TValue *` that points to the upvalue array of the closure.

I suspect that if we use `TValue *` the generated C code will look a bit prettier and it might slightly help with the `restrict` stuff. However, I expect that the performance would be similar either way.

# Restrict

We should consider declaring the `base`, `K` and `U` as restrict. Perhaps it might help the compiler optimize things, by letting it know that the contents of these arrays will not be changed if there is an intervening function call. That is, if we fetch one of the upvalues, we won't need to fetch it again later.

In my simplistic tests on Godbolt, Clang could optimize it but GCC did not: https://godbolt.org/z/zPhTja9sG

```c
StackValue * restrict base,
TValue * restrict K,
TValue * restrict U,
```

# The constant table

Because functions might potentially contain a large number of string literals (more than 255), it is not possible to store the string literals directly as closure upvalues. Therefore, we must have a separate object for the constant table.

We can keep using an userdata for the constant table, which has a limit of 2^16 uservalues. If there is a need for more constants than this, we can come up with a workaround.

In theory, we could have a separate constant table for each function, similarly to what the Lua closures do. However, I think it would be acceptable to have all the closures in a module share the same constant table.

# The K argument

Just as today, the first (zero-th) upvalue of every closure should be the constant table. This begs the question of whether the Pallene entry point should continue to receive the `K` as a separate argument, or if it should receive only the `U` as an argument.

We need to run some experiments to decide. But until then, my gut instinct is to have a separate argument for the `K`.

**Comment by srijan-paul** [comment]
Would it help to tackle this in two parts?
First we could do the easier part where upvalues are passed as `TValue*` to the Pallene entry point, and then sort out the `G`, which may require several changes from the checker and down.

**Comment by hugomg** [comment]
I expect that this issue might require several PRs.

**Comment by hugomg** [comment]
Fixed in #462. Yay!
